File: native/src/bridge.cpp
#include "bridge.h"

#include "common.h"
#include "fsapi.h"
#include "netapi.h"
#include "sysapi.h"

namespace otz {
namespace {

// משתנה סביבה, או ריק כשאינו מוגדר. הזיהוי של אוצריא נשען עליהם
// (`%LOCALAPPDATA%\Programs\Otzaria` וכו'), וכך הם מגיעים ל-JS במקום
// אחד ולא בקריאה נפרדת לכל אחד.
std::wstring EnvironmentValue(const wchar_t* name) {
  const DWORD needed = GetEnvironmentVariableW(name, nullptr, 0);
  if (needed == 0) return {};
  std::wstring value(needed, L'\0');
  const DWORD written = GetEnvironmentVariableW(name, value.data(), needed);
  if (written == 0) return {};
  value.resize(written);
  return value;
}

// ארגומנט לפי אינדקס, או ריק כשלא נמסר. כל הפקודות מקבלות את
// הארגומנטים שלהן דרך זה, כדי שהודעה קטועה תיתן שדה ריק ולא קריסה.
std::wstring Arg(const std::vector<std::wstring>& fields, size_t index) {
  // fields[0] הוא ה-reqId ו-fields[1] הפקודה; הארגומנטים מתחילים ב-2.
  const size_t at = index + 2;
  return at < fields.size() ? fields[at] : std::wstring{};
}

long long ArgNumber(const std::vector<std::wstring>& fields, size_t index,
                    long long fallback) {
  const std::wstring text = Arg(fields, index);
  if (text.empty()) return fallback;
  try {
    return std::stoll(text);
  } catch (...) {
    // שדה שאינו מספר — נופלים לברירת המחדל ולא מפילים את הבקשה.
    return fallback;
  }
}

}  // namespace

// ── WorkerPool ───────────────────────────────────────────────────────────────

void WorkerPool::Start(size_t workers) {
  for (size_t i = 0; i < workers; ++i) {
    threads_.emplace_back([this] { Run(); });
  }
}

void WorkerPool::Stop() {
  {
    std::lock_guard<std::mutex> guard(mutex_);
    if (stopping_) return;
    stopping_ = true;
  }
  ready_.notify_all();
  for (auto& thread : threads_) {
    if (thread.joinable()) thread.join();
  }
  threads_.clear();
}

void WorkerPool::Post(std::function<void()> job) {
  {
    std::lock_guard<std::mutex> guard(mutex_);
    if (stopping_) return;
    jobs_.push_back(std::move(job));
  }
  ready_.notify_one();
}

void WorkerPool::Run() {
  while (true) {
    std::function<void()> job;
    {
      std::unique_lock<std::mutex> lock(mutex_);
      ready_.wait(lock, [this] { return stopping_ || !jobs_.empty(); });
      // עבודה שנשארה בתור בזמן כיבוי אינה מתחילה — בדיוק כמו בדיקת
      // הביטול ב-`runPooled`.
      if (stopping_) return;
      job = std::move(jobs_.front());
      jobs_.pop_front();
    }
    // חריג מעבודה בודדת לא מפיל את ה-thread ואיתו את כל התור.
    try {
      job();
    } catch (...) {
      LogError(L"עבודה ברקע נפלה בחריג");
    }
  }
}

// ── Bridge ───────────────────────────────────────────────────────────────────

Bridge::Bridge(AppPaths paths, std::wstring version)
    : paths_(std::move(paths)), version_(std::move(version)) {
  pool_.Start();
}

Bridge::~Bridge() {
  closing_ = true;
  pool_.Stop();
}

void Bridge::Attach(UiDispatcher ui, JsSender send_to_js, WindowCommand window) {
  ui_ = std::move(ui);
  send_to_js_ = std::move(send_to_js);
  window_ = std::move(window);
}

void Bridge::ReplyOk(long long request_id, const std::string& json) {
  if (closing_ || !send_to_js_) return;
  send_to_js_(JsonObject({
      {"id", JsonNumber(request_id)},
      {"ok", "true"},
      {"result", json},
  }));
}

void Bridge::ReplyError(long long request_id, const std::wstring& error) {
  if (closing_ || !send_to_js_) return;
  send_to_js_(JsonObject({
      {"id", JsonNumber(request_id)},
      {"ok", "false"},
      {"error", JsonString(error)},
  }));
}

void Bridge::RunAsync(long long request_id, std::function<Result()> job) {
  pool_.Post([this, request_id, job = std::move(job)] {
    Result result = job();
    if (closing_) return;
    // התשובה חוזרת ל-thread של הממשק לפני שהיא נוגעת ב-WebView2.
    ui_([this, request_id, result = std::move(result)] {
      if (result.ok) {
        ReplyOk(request_id, result.json);
      } else {
        ReplyError(request_id, result.error);
      }
    });
  });
}

void Bridge::HandleMessage(const std::wstring& message) {
  const std::vector<std::wstring> fields = SplitFields(message);
  if (fields.size() < 2) {
    LogError(L"הודעה פגומה מהממשק");
    return;
  }

  long long request_id = 0;
  try {
    request_id = std::stoll(fields[0]);
  } catch (...) {
    LogError(L"מזהה בקשה שאינו מספר: " + fields[0]);
    return;
  }
  const std::wstring& command = fields[1];

  // ── קבצים ─────────────────────────────────────────────────────────────────
  if (command == L"fs.readText") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::ReadText(path); });
    return;
  }
  if (command == L"fs.writeText") {
    const std::wstring path = Arg(fields, 0);
    // הטקסט הוא השדה השני, וייתכן שהוא מאות KB (הקטלוג). הוא אינו
    // יכול להכיל את מפריד השדות, ולכן הוא עובר שלם.
    const std::string text = Utf8(Arg(fields, 1));
    RunAsync(request_id, [path, text] {
      return fsapi::WriteTextAtomic(path, text);
    });
    return;
  }
  if (command == L"fs.kind") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::PathKind(path); });
    return;
  }
  if (command == L"fs.stat") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::Stat(path); });
    return;
  }
  if (command == L"fs.list") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::ListDir(path); });
    return;
  }
  if (command == L"fs.mkdirs") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::MakeDirs(path); });
    return;
  }
  if (command == L"fs.delete") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id, [path] { return fsapi::DeleteFileAt(path); });
    return;
  }
  if (command == L"fs.copy") {
    const std::wstring from = Arg(fields, 0);
    const std::wstring to = Arg(fields, 1);
    RunAsync(request_id, [from, to] { return fsapi::CopyFileTo(from, to); });
    return;
  }
  if (command == L"fs.rename") {
    const std::wstring from = Arg(fields, 0);
    const std::wstring to = Arg(fields, 1);
    RunAsync(request_id, [from, to] { return fsapi::Rename(from, to); });
    return;
  }
  if (command == L"fs.readBase64") {
    const std::wstring path = Arg(fields, 0);
    const auto offset = static_cast<uint64_t>(ArgNumber(fields, 1, 0));
    const auto length = static_cast<uint32_t>(ArgNumber(fields, 2, 0));
    RunAsync(request_id, [path, offset, length] {
      return fsapi::ReadBase64(path, offset, length);
    });
    return;
  }

  // ── רשת ───────────────────────────────────────────────────────────────────
  if (command == L"net.get") {
    const std::wstring url = Arg(fields, 0);
    const auto timeout =
        static_cast<int>(ArgNumber(fields, 1, netapi::kDefaultTimeoutMs));
    RunAsync(request_id, [url, timeout] { return netapi::Get(url, timeout); });
    return;
  }
  if (command == L"net.download") {
    const std::wstring url = Arg(fields, 0);
    const std::wstring dest = Arg(fields, 1);
    const auto timeout =
        static_cast<int>(ArgNumber(fields, 2, netapi::kDefaultTimeoutMs));
    const auto stall =
        static_cast<int>(ArgNumber(fields, 3, netapi::kDefaultStallMs));
    RunAsync(request_id, [url, dest, timeout, stall] {
      return netapi::Download(url, dest, timeout, stall);
    });
    return;
  }

  // ── מערכת ─────────────────────────────────────────────────────────────────
  if (command == L"sys.openUrl") {
    // ShellExecute רץ על ה-thread של הממשק: הוא COM/STA, והוא מהיר.
    const std::wstring url = Arg(fields, 0);
    const Result result = sysapi::OpenExternalUrl(url);
    if (result.ok) {
      ReplyOk(request_id, result.json);
    } else {
      ReplyError(request_id, result.error);
    }
    return;
  }
  if (command == L"sys.launch") {
    const std::wstring exe = Arg(fields, 0);
    const std::wstring argument = Arg(fields, 1);
    RunAsync(request_id, [exe, argument] {
      return sysapi::LaunchDetached(exe, argument);
    });
    return;
  }
  if (command == L"sys.exeVersion") {
    const std::wstring path = Arg(fields, 0);
    RunAsync(request_id,
             [path] { return sysapi::ReadExeProductVersion(path); });
    return;
  }
  if (command == L"sys.registryDirs") {
    // סריקת הרג'יסטרי לוקחת ~200ms — לא על ה-thread של הממשק.
    RunAsync(request_id, [] { return sysapi::OtzariaRegistryInstallDirs(); });
    return;
  }
  if (command == L"sys.runningOtzaria") {
    RunAsync(request_id, [] { return sysapi::FindRunningOtzaria(); });
    return;
  }

  // ── התוכנה עצמה ───────────────────────────────────────────────────────────
  if (command == L"app.info") {
    ReplyOk(request_id, JsonObject({
                            {"version", JsonString(version_)},
                            {"dataDir", JsonString(paths_.data_dir)},
                            {"pluginsDir", JsonString(paths_.PluginsDir())},
                            {"catalogPath", JsonString(paths_.CatalogPath())},
                            {"stateDir", JsonString(paths_.state_dir)},
                            {"stateFile", JsonString(paths_.StateFilePath())},
                            {"exePath", JsonString(ExecutablePath())},
                            {"readOnly", JsonBool(paths_.read_only)},
                            {"darkMode", JsonBool(sysapi::IsSystemDarkMode())},
                            // ריק = הכול תקין. כשיש כאן טקסט, הממשק מציג
                            // אותו כמסך שגיאה ואינו טוען את החנות — פורט
                            // של `SetupErrorScreen`.
                            {"setupError", JsonString(paths_.error)},
                            // המיקומים שהזיהוי של אוצריא סורק.
                            {"env", JsonObject({
                                        {"localAppData",
                                         JsonString(EnvironmentValue(L"LOCALAPPDATA"))},
                                        {"appData",
                                         JsonString(EnvironmentValue(L"APPDATA"))},
                                        {"programFiles",
                                         JsonString(EnvironmentValue(L"ProgramFiles"))},
                                        {"programData",
                                         JsonString(EnvironmentValue(L"ProgramData"))},
                                    })},
                        }));
    return;
  }

  if (command == L"app.log") {
    // ⚠️ היומן הוא **הדרך היחידה** לדעת מה קרה אצל המשתמש: הממשק רץ
    // בתוך WebView2, ושגיאת JS שאין לה יעד נעלמת עם החלון. לכן גם
    // `window.onerror` נמסר לכאן.
    const std::wstring level = Arg(fields, 0);
    const std::wstring text = Arg(fields, 1);
    LogLine(level.empty() ? L"INFO" : level, text);
    ReplyOk(request_id, "true");
    return;
  }

  // ── דיאלוג ופעולות חלון ───────────────────────────────────────────────────
  // כולן מחייבות את ה-thread של הממשק ואת ה-HWND, ולכן מופנות ל-main.cpp.
  if (command == L"sys.saveDialog" || command.rfind(L"win.", 0) == 0) {
    if (window_) {
      window_(command, std::vector<std::wstring>(fields.begin() + 2, fields.end()),
              request_id);
    } else {
      ReplyError(request_id, L"החלון אינו זמין");
    }
    return;
  }

  ReplyError(request_id, L"פקודה שאינה מוכרת: " + command);
}

}  // namespace otz

File: native/web/js/bridge.js
// הצד ה-JS של הגשר אל ה-host.
//
// ── המסגור ───────────────────────────────────────────────────────────────────
//   אלינו → host   `<reqId>\x1f<command>\x1f<arg0>\x1f<arg1>…`
//   host → אלינו   `{"id":…,"ok":true,"result":…}` / `{…,"ok":false,"error":…}`
//                  או `{"event":"<name>", …}` לאירוע שאינו תשובה
//
// מפריד השדות הוא U+001F, שאינו חוקי בנתיב ואינו מופיע בכתובת — ולכן
// אין כאן escaping ואין מה לפרסר בצד ה-C++. ראו ההסבר ב-native/src/bridge.h.

const FIELD_SEP = '\x1f';

/** בקשות שממתינות לתשובה: reqId → {resolve, reject, watchdog}. */
const pending = new Map();

/**
 * הפקודות שאין להן גבול זמן סביר, ולכן **אין עליהן שומר**.
 *
 * הורדה של תוסף היא עשרות MB בקו איטי, העתקה היא כתיבה לכונן נייד
 * שעלול להיות אטי, וחלון השמירה פתוח עד שהמשתמש בוחר. לכל אלה אין מספר
 * שאפשר להגיד עליו "מכאן זה תקוע" — ושומר שהיה יורה עליהן היה **גרוע
 * מהבאג**: הוא אינו עוצר את ה-host, שממשיך לעבוד, ולכן כל מה שהיה משיג
 * הוא ניתוק בין שני הצדדים (הממשק אומר "נכשל" בעוד הקובץ נכתב).
 */
const UNBOUNDED_COMMANDS = new Set([
  'net.get', 'net.download', 'fs.copy', 'sys.saveDialog',
]);

/**
 * שומר לשאר הפקודות. **אינו פסק זמן של הממשק אלא רשת ביטחון**: כל פקודה
 * שה-host מכיר משיבה בדיוק פעם אחת, וגם פקודה שאינה מוכרת מקבלת שגיאה —
 * ולכן אם עברו עשר דקות בלי תשובה משהו נבלע בדרך. בלי זה הבקשה נשארת
 * תלויה לנצח, והמסך נתקע ב"טוען" בלי שגיאה ובלי כפתור לנסות שוב.
 */
const WATCHDOG_MS = 10 * 60 * 1000;

/** מאזינים לאירועים מה-host: שם → קבוצת פונקציות. */
const listeners = new Map();

let nextRequestId = 1;

/** `true` כשאנחנו רצים בתוך ה-host ולא בדפדפן רגיל. */
export const hasHost = typeof chrome !== 'undefined' &&
    chrome.webview !== undefined && chrome.webview !== null;

/**
 * סוגר בקשה ממתינה: מוציא אותה מהמפה, מכבה את השומר, ומריץ עליה
 * [apply]. מחזיר `false` אם כבר נסגרה (תשובה שהגיעה פעמיים, או אחרי
 * שהשומר ירה).
 */
function settle(id, apply) {
  const request = pending.get(id);
  if (request === undefined) return false;
  pending.delete(id);
  if (request.watchdog !== null) clearTimeout(request.watchdog);
  apply(request);
  return true;
}

/**
 * מחלץ את מזהה הבקשה מתשובה שלא ניתן היה לפרסר.
 *
 * ה-host כותב את `id` **ראשון** (ראו `Bridge::ReplyOk`), ולכן הוא קריא
 * גם כשהמשך המחרוזת פגום — וזה מה שמאפשר לשחרר את מי שממתין לה במקום
 * להשאיר אותו תלוי.
 */
function requestIdOf(text) {
  if (typeof text !== 'string') return null;
  const match = /^\s*\{\s*"id"\s*:\s*(\d+)/.exec(text);
  return match === null ? null : Number(match[1]);
}

if (hasHost) {
  chrome.webview.addEventListener('message', (raw) => {
    let message;
    try {
      // ה-host מוסר מחרוזת (`PostWebMessageAsString`), ולכן `data` היא
      // המחרוזת עצמה ולא אובייקט.
      message = JSON.parse(raw.data);
    } catch (error) {
      console.error('הודעה פגומה מה-host:', raw.data, error);
      // תשובה פגומה אינה מגיעה לאיש — ובלי לשחרר את הבקשה שהיא נועדה
      // לה, היא הייתה נשארת ממתינה עד סוף ההרצה.
      const id = requestIdOf(raw.data);
      if (id !== null) {
        settle(id, (request) => request.reject(
            new HostError('תשובה פגומה מה-host')));
      }
      return;
    }

    if (message.event !== undefined) {
      const handlers = listeners.get(message.event);
      if (handlers) {
        for (const handler of handlers) {
          // מאזין שנפל לא מונע מהשאר לרוץ.
          try {
            handler(message);
          } catch (error) {
            console.error(`מאזין ל-${message.event} נפל:`, error);
          }
        }
      }
      return;
    }

    settle(message.id, (request) => {
      if (message.ok) {
        request.resolve(message.result);
      } else {
        request.reject(new HostError(message.error));
      }
    });
  });
}

/**
 * שגיאה שהגיעה מה-host. טיפוס נפרד כדי שהממשק יוכל להבחין בין "מערכת
 * ההפעלה אמרה לא" לבין באג אצלנו — הראשון מוצג למשתמש כמו שהוא.
 */
export class HostError extends Error {
  constructor(message) {
    super(message);
    this.name = 'HostError';
  }
}

/**
 * שולח פקודה ל-host ומחזיר Promise לתשובה.
 *
 * הארגומנטים מומרים למחרוזות; `null`/`undefined` הופכים למחרוזת ריקה,
 * שהיא ארגומנט לגיטימי בצד ה-C++ (למשל סיומת מועדפת שלא נמסרה).
 */
export function call(command, ...args) {
  if (!hasHost) {
    return Promise.reject(
        new HostError(`אין host — הפקודה ${command} אינה זמינה`));
  }

  const id = nextRequestId++;
  let request;
  const promise = new Promise((resolve, reject) => {
    request = {resolve, reject, watchdog: null};
  });
  pending.set(id, request);
  if (!UNBOUNDED_COMMANDS.has(command)) {
    request.watchdog = setTimeout(() => {
      settle(id, (waiting) => waiting.reject(
          new HostError(`ה-host לא השיב על ${command}`)));
    }, WATCHDOG_MS);
  }

  const fields = [String(id), command];
  for (const arg of args) {
    fields.push(arg === null || arg === undefined ? '' : String(arg));
  }
  try {
    chrome.webview.postMessage(fields.join(FIELD_SEP));
  } catch (error) {
    // ההודעה לא יצאה, ולכן אין מי שישיב לה. סוגרים כאן ולא משאירים
    // בקשה שממתינה לתשובה שלא תגיע לעולם.
    settle(id, (waiting) => waiting.reject(new HostError(
        `שליחת הפקודה ${command} נכשלה: ${error?.message ?? error}`)));
  }
  return promise;
}

/** נרשם לאירוע מה-host. מחזיר פונקציה שמבטלת את ההרשמה. */
export function on(eventName, handler) {
  let handlers = listeners.get(eventName);
  if (handlers === undefined) {
    handlers = new Set();
    listeners.set(eventName, handlers);
  }
  handlers.add(handler);
  return () => handlers.delete(handler);
}

// ── עטיפות נוחות ─────────────────────────────────────────────────────────────
// שמות מפורשים במקום מחרוזות פקודה פזורות בכל הקוד. כל אחת מהן היא
// שורה אחת, וכולן יחד הן ה-API שהשאר משתמש בו.

export const fs = {
  readText: (path) => call('fs.readText', path),
  writeText: (path, text) => call('fs.writeText', path, text),
  kind: (path) => call('fs.kind', path),
  stat: (path) => call('fs.stat', path),
  list: (path) => call('fs.list', path),
  mkdirs: (path) => call('fs.mkdirs', path),
  remove: (path) => call('fs.delete', path),
  // ⚠️ **ריקה בלבד.** תיקייה שנשאר בה תוכן חוזרת `false` ואינה נמחקת —
  // ראו `RemoveEmptyDirAt` ב-fsapi.cpp.
  removeDir: (path) => call('fs.removeDir', path),
  copy: (from, to) => call('fs.copy', from, to),
  rename: (from, to) => call('fs.rename', from, to),
  readBase64: (path, offset, length) =>
      call('fs.readBase64', path, offset, length),

  /** `true` אם הנתיב הוא קובץ קיים. */
  async fileExists(path) {
    return (await this.kind(path)) === 'file';
  },

  /** `true` אם הנתיב הוא תיקייה קיימת. */
  async dirExists(path) {
    return (await this.kind(path)) === 'dir';
  },
};

export const net = {
  get: (url, timeoutMs) => call('net.get', url, timeoutMs),
  download: (url, destPath, timeoutMs, stallMs) =>
      call('net.download', url, destPath, timeoutMs, stallMs),
};

export const sys = {
  openUrl: (url) => call('sys.openUrl', url),
  launch: (exePath, argument) => call('sys.launch', exePath, argument),
  saveDialog: (suggestedName) => call('sys.saveDialog', suggestedName),
  exeVersion: (path) => call('sys.exeVersion', path),
  registryDirs: () => call('sys.registryDirs'),
  runningOtzaria: () => call('sys.runningOtzaria'),
};

export const win = {
  minimize: () => call('win.minimize'),
  maximizeToggle: () => call('win.maximizeToggle'),
  close: () => call('win.close'),
  state: () => call('win.state'),
  dragMove: () => call('win.dragMove'),
  resizeStart: (edge) => call('win.resizeStart', edge),
};

/** פרטי התוכנה והנתיבים. נקרא פעם אחת בעלייה. */
export const appInfo = () => call('app.info');

/**
 * כותב ליומן שה-host מחזיק.
 *
 * ⚠️ זה **הדרך היחידה** לדעת מה קרה אצל המשתמש: הממשק רץ בתוך
 * WebView2, ו-`console.log` נעלם עם החלון. `main.js` מפנה לכאן גם את
 * `window.onerror` ואת ה-promise שנדחו בלי מטפל.
 */
export function logToFile(level, message) {
  if (!hasHost) return Promise.resolve();
  // כשל בכתיבת יומן אינו סיבה להפיל את מה שקרה סביבו.
  return call('app.log', level, message).catch(() => {});
}
